### DISCUSSION TITLE
Limitation on size of socket.send()

### DISCUSSION THREAD

**Discussion by BMOC1** [micropython/micropython#11568] [Using MicroPython] [answered]
I'm building an html response.  It appears i have hit a limitation of right around 6300 bytes.
I am successfully sending a Response Length:6327

I've noticed if i go much over that, it starts truncating out on the webbrowser/client side.

```
cl, addr = s.accept()
request = cl.recv(1024)
~~~~  ~~~~~ ~~~~~  (additional code not relevant)
cl.send('HTTP/1.0 200 OK\r\nContent-type: text/html\r\n\r\n')
print("Response Length:" + str(len(response)))
cl.send(response)
```
here is my software/hardware:
3.4.0: MicroPython v1.20.0 on 2023-04-26
Raspberry Pi Pico W with RP2040

I do not have enough knowledge to know where the limiting factor is coming from.  Is it a limit of the actual string variable? Actually, i just thought of that, ill do a test to see if the variable has the full data.  Do i break it up, and send it in multiple send commands?  - Should i be sending it in a different way?

I appreciate any input.. Thank you.

**Comment by BMOC1** [comment]
**EDIT - This solution proved NOT to work.**

I believe that the .send is probably not parsing the data into manageable packet sizes or a buffer limit is being exceeded

I figured out a solution...   Hopefully my method is clean....   I broke the string variable into an array of chunk sizes.
In this test, i appended a bunch of data into the response string variable making it 3x the size that failed before with a [END] marker, and ran the code.  -  I was able to view the entire HTML document with the visible [END] marker proving it was sending/receiving the entirety of the file.

```
print("Response Length:" + str(len(response)))  #DEBUG LINE
chunksize = 1024
chunkarray = [response[i:i+chunksize] for i in range(0,len(response)-1,chunksize)]
for chunk in chunkarray:
    cl.send(chunk)
    print(len(chunk))  #DEBUG LINE
cl.close()
```

```
Response Length:20653
1024
1024
1024
1024
1024
1024
1024
1024
1024
1024
1024
1024
1024
1024
1024
1024
1024
1024
1024
1024
173
```

**Reply by robert-hh** [reply, marked as answer]
Socket.send() is not guaranteed to send all data. You have to check the return value. Or use socket.sendall().

**Reply by BMOC1** [reply]
Thank YOU.  I just realized this, as the same problem just crept back up... It's cutting text, and its VERY RANDOM.

**Reply by BMOC1** [reply]
I switched from **.send()** to **.sendall()** reverting back to my original code of not breaking it up into chunk sizes, and it **worked**.  When i read the MicroPython documentation on **sendall()** , it suggested the use of the **write()** method instead of **sendall()**. 

> The behaviour of this method on non-blocking sockets is undefined. Due to this, on MicroPython, it’s recommended to use [write()](https://docs.micropython.org/en/latest/library/socket.html#socket.socket.write) method instead, which has the same “no short writes” policy for blocking sockets, and will return number of bytes sent on non-blocking sockets.

I am currently using blocking sockets, although, I'm gonna have to open up another discussion on that point, as i am utilizing 
`import uselect as select` on socket to act asynchronously so that i can process other tasks waiting for a connection.

So, should i use **.write()** command instead of **.sendall()**  - If i understand the documentation correctly, utilizing blocking sockets, it doesn't matter as they act the same, but if i switch to unblocking, then I should definitely use the **write()**?

**Reply by robert-hh** [reply]
As of the text you cited, using socket.write() will be no advantage in the case of non-blocking sockets. write() may then send partial data just like send(), and you have to ensure data completeness in your code.
